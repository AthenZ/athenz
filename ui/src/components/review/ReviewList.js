/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import styled from '@emotion/styled';
import GroupReviewTable from '../group/GroupReviewTable';
import ReviewTable from './ReviewTable';
import Alert from '../denali/Alert';
import { MODAL_TIME_OUT } from '../constants/constants';
import { selectIsLoading } from '../../redux/selectors/loading';
import { selectTimeZone } from '../../redux/selectors/domains';
import { connect } from 'react-redux';
import { ReduxPageLoader } from '../denali/ReduxPageLoader';
import { withRouter } from 'next/router';

const RolesSectionDiv = styled.div`
    margin: 20px;
    margin-bottom: 0px;
`;

// dont need to make it as redux because it get props from groups and role and in order to not need to figure out
// which data to get from the store it easier to get the data from the father component which is seperated between roles and groups
class ReviewList extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            showuser: false,
            members: props.members || [],
            errorMessage: null,
        };
        this.closeModal = this.closeModal.bind(this);
        this.submitSuccess = this.submitSuccess.bind(this);
    }

    componentDidUpdate = (prevProps) => {
        if (
            prevProps.collection !== this.props.collection ||
            prevProps.domain !== this.props.domain
        ) {
            this.setState({
                members: this.props.members,
                showuser: false,
            });
        }
    };

    submitSuccess(successMessage) {
        this.setState({
            showSuccess: true,
            successMessage,
            errorMessage: null,
        });
        this.props.onSuccessReview &&
            this.props.onSuccessReview(successMessage + ` Removed ${this.props.category} from view.`);
        setTimeout(() => {
            this.setState({
                showSuccess: false,
            });
            if (!this.props.isCardView) {
                this.props.router.push(
                    `/domain/${this.props.domain}/${this.props.category}/${this.props.collection}/members`,
                    `/domain/${this.props.domain}/${this.props.category}/${this.props.collection}/members`
                );
            }
        }, MODAL_TIME_OUT);
    }

    closeModal() {
        this.setState({ showSuccess: null });
    }

    render() {
        const { domain, collection, collectionDetails } = this.props;
        return this.props.isLoading.length !== 0 ? (
            <ReduxPageLoader message={'Loading data'} />
        ) : (
            <RolesSectionDiv data-testid='review-list'>
                {this.props.category === 'group' && (
                    <GroupReviewTable
                        domain={domain}
                        groupName={collection}
                        timeZone={this.props.timeZone}
                        _csrf={this.props._csrf}
                        onUpdateSuccess={this.submitSuccess}
                        justification={this.props.justification}
                    />
                )}
                {this.props.category === 'role' && (
                    <ReviewTable
                        domain={domain}
                        role={collection}
                        roleDetails={collectionDetails}
                        members={this.props.members}
                        timeZone={this.props.timeZone}
                        _csrf={this.props._csrf}
                        onUpdateSuccess={this.submitSuccess}
                        justification={this.props.justification}
                    />
                )}
                {this.state.showSuccess ? (
                    <Alert
                        isOpen={this.state.showSuccess}
                        title={this.state.successMessage}
                        onClose={this.closeModal}
                        type='success'
                    />
                ) : null}
            </RolesSectionDiv>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        ...props,
        isLoading: selectIsLoading(state),
        timeZone: selectTimeZone(state),
    };
};

export default connect(mapStateToProps)(withRouter(ReviewList));
