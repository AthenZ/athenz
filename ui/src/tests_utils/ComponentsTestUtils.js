/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { initStore } from '../redux/store';
import { getExpiryTime } from '../redux/utils';

export const renderWithRedux = (component, initialState = {}) => {
    return render(
        <Provider store={initStore(initialState)}>{component}</Provider>
    );
};

export const buildUserForState = (user) => {
    return {
        expiry: getExpiryTime(),
        user,
    };
};

export const getStateWithUser = (user, initialState = {}) => {
    return { ...initialState, user };
};

export const buildDomainDataForState = (domainData, domainName = 'dom') => {
    return {
        domainName,
        expiry: getExpiryTime(),
        domainData,
    };
};

export const getStateWithDomainData = (domainData, initialState = {}) => {
    return { ...initialState, domainData: domainData };
};

export const getStateWithRoles = (roles, initialState = {}) => {
    return { ...initialState, roles };
};

export const buildRolesForState = (roles, domainName = 'dom') => {
    return {
        domainName,
        expiry: getExpiryTime(),
        roles,
    };
};

export const buildPoliciesForState = (policies, domainName = 'dom') => {
    return {
        domainName,
        expiry: getExpiryTime(),
        policies,
    };
};

export const getStateWithPolicies = (policies, initialState = {}) => {
    return { ...initialState, policies };
};

export const buildServicesForState = (
    services,
    domainName = 'dom',
    allProviders = []
) => {
    return {
        domainName,
        expiry: getExpiryTime(),
        services,
        allProviders,
    };
};

export const getStateWithServices = (services, initialState = {}) => {
    return { ...initialState, services };
};

export const buildMicrosegmentationForState = (
    inboundOutboundList,
    domainName = 'dom'
) => {
    return {
        domainName,
        expiry: getExpiryTime(),
        inboundOutboundList,
    };
};

export const getStateWithMicrosegmentation = (
    microsegmentation,
    initialState = {}
) => {
    return { ...initialState, microsegmentation };
};

export const buildGroupsForState = (groups, domainName = 'dom') => {
    return {
        domainName,
        expiry: getExpiryTime(),
        groups,
    };
};

export const getStateWithGroups = (groups, initialState = {}) => {
    return { ...initialState, groups };
};

export const buildServiceDependencies = (
    serviceDependencies,
    domainName = 'dom'
) => {
    return {
        domainName,
        expiry: getExpiryTime(),
        serviceDependencies,
    };
};

export const getStateWithServiceDependencies = (
    serviceDependencies,
    initialState = {}
) => {
    return { ...initialState, serviceDependencies: serviceDependencies };
};

export const mockRolesApiCalls = () => {
    return {
        getRoles: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve([]);
            })
        ),
        getRoleMembers: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve({});
            })
        ),
    };
};

export const mockPoliciesApiCalls = () => {
    return {
        getPolicies: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve([]);
            })
        ),
        getPolicy: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve({});
            })
        ),
        // getRoleMembers: jest.fn().mockReturnValue(
        //     new Promise((resolve, reject) => {
        //         resolve({});
        //     })
        // ),
    };
};

export const mockAllDomainDataApiCalls = (domainDetails, headerDetails) => {
    return {
        getDomain: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve(domainDetails || {});
            })
        ),
        isAWSTemplateApplied: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve('');
            })
        ),
        getHeaderDetails: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve(headerDetails || {});
            })
        ),
        getPendingDomainMembersListByDomain: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve([]);
            })
        ),
        getFeatureFlag: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve('');
            })
        ),
        getMeta: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve([]);
            })
        ),
        getAuthorityAttributes: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve({});
            })
        ),
        getReviewGroups: jest.fn().mockReturnValue([]),
        getReviewRoles: jest.fn().mockReturnValue([]),
        getPageFeatureFlag: jest.fn().mockResolvedValue({}),
    };
};

export const getStateWithUserList = (userList, initialState = {}) => {
    return { ...initialState, user: userList };
};

export const getStateWithDomainDataAndUserList = (
    domainData,
    userList,
    initialState = {}
) => {
    return { ...initialState, domainData: domainData, user: userList };
};
