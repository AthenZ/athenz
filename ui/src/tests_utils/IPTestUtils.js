/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export default class IPTestUtils {
    static getRandomInt(min, max) {
        return Math.floor(Math.random() * (max - min + 1)) + min;
    }

    static generateRandomIP() {
        return `${IPTestUtils.getRandomInt(0, 255)}.${IPTestUtils.getRandomInt(
            0,
            255
        )}.${IPTestUtils.getRandomInt(0, 255)}.${IPTestUtils.getRandomInt(
            0,
            255
        )}`;
    }

    static generateRandomCIDR() {
        const ip = IPTestUtils.generateRandomIP();
        const prefix = IPTestUtils.getRandomInt(0, 32);
        return `${ip}/${prefix}`;
    }

    static generateRandomString(length) {
        const characters =
            'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        let result = '';
        for (let i = 0; i < length; i++) {
            result += characters.charAt(
                Math.floor(Math.random() * characters.length)
            );
        }
        return result;
    }

    static generateRandomFQDN() {
        const subdomain = IPTestUtils.generateRandomString(
            IPTestUtils.getRandomInt(3, 10)
        );
        const domain = IPTestUtils.generateRandomString(
            IPTestUtils.getRandomInt(3, 10)
        );
        const tld = IPTestUtils.generateRandomString(
            IPTestUtils.getRandomInt(2, 6)
        ); // Top Level Domain, like com, net, org, etc.
        return `${subdomain}.${domain}.${tld}`;
    }
}
