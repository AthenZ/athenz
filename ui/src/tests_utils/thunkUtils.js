/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import {
    addAssertionPolicyVersionToStore,
    deleteAssertionPolicyVersionFromStore,
    loadPolicies,
} from '../redux/actions/policies';
import { getExpiryTime } from '../redux/utils';
import { storeDomainData, storePolicies } from '../redux/actions/domains';
import { loadDomainData } from '../redux/actions/domain-data';

export const getLoadDomainDataAction = (domainName, domainData) => {
    return loadDomainData(domainData, domainName, getExpiryTime());
};

export const getStoreDomainDataAction = (domainData) => {
    return storeDomainData(domainData);
};

export const getLoadPoliciesAction = (domainName, policies = {}) => {
    return loadPolicies(policies, domainName, getExpiryTime());
};

export const getStorePoliciesAction = (policiesData) => {
    return storePolicies(policiesData);
};

export const getDeleteAssertionPolicyVersionAction = (
    policyName,
    version,
    assertionId
) => {
    return deleteAssertionPolicyVersionFromStore(
        policyName,
        version,
        assertionId
    );
};

export const getAddAssertionPolicyVersionAction = (
    policyName,
    version,
    newAssertion
) => {
    return addAssertionPolicyVersionToStore(policyName, version, newAssertion);
};
