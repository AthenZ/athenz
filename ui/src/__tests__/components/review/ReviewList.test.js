/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { render } from '@testing-library/react';
import ReviewList from '../../../components/review/ReviewList';
import API from '../../../api';
import { renderWithRedux } from '../../../tests_utils/ComponentsTestUtils';

describe('ReviewList', () => {
    it('should render', () => {
        let members = [];
        let domain = 'domain';
        let role = 'roleName';
        const roleDetails = {
            trust: null,
            reviewEnabled: true,
            selfServe: true,
        };
        let user1 = {
            memberName: 'user1',
            approved: true,
        };
        let user2 = {
            memberName: 'user2',
            approved: false,
        };
        let user3 = {
            memberName: 'user3',
            approved: false,
        };
        let user4 = {
            memberName: 'user4',
            approved: true,
        };
        members.push(user1);
        members.push(user2);
        members.push(user3);
        members.push(user4);

        const { getByTestId } = renderWithRedux(
            <ReviewList
                domain={domain}
                role={role}
                roleDetails={roleDetails}
                members={members}
                isDomainAuditEnabled={true}
            />
        );
        const reviewlist = getByTestId('review-list');

        expect(reviewlist).toMatchSnapshot();
    });
    it('should render delegated role', () => {
        let members = [];
        let domain = 'domain';
        let role = 'roleName';
        const roleDetails = {
            trust: 'domain.delegated',
            reviewEnabled: false,
            selfServe: true,
        };
        let user1 = {
            memberName: 'user1',
            approved: true,
        };
        let user2 = {
            memberName: 'user2',
            approved: false,
        };
        let user3 = {
            memberName: 'user3',
            approved: false,
        };
        let user4 = {
            memberName: 'user4',
            approved: true,
        };
        members.push(user1);
        members.push(user2);
        members.push(user3);
        members.push(user4);

        const { getByTestId } = renderWithRedux(
            <ReviewList
                domain={domain}
                role={role}
                roleDetails={roleDetails}
                members={members}
                isDomainAuditEnabled={true}
            />
        );
        const reviewlist = getByTestId('review-list');

        expect(reviewlist).toMatchSnapshot();
    });
});
