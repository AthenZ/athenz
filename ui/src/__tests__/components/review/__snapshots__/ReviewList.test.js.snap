// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewList should render 1`] = `
.emotion-0 {
  margin: 20px;
  margin-bottom: 0px;
}

<div
  class="emotion-0 emotion-1"
  data-testid="review-list"
/>
`;

exports[`ReviewList should render delegated role 1`] = `
.emotion-0 {
  margin: 20px;
  margin-bottom: 0px;
}

<div
  class="emotion-0 emotion-1"
  data-testid="review-list"
/>
`;
